File: src/main/java/edu/cornell/cs/apl/viaduct/imp/parsing/SourceRange.java
package edu.cornell.cs.apl.viaduct.imp.parsing;

import com.google.auto.value.AutoValue;
import java.io.PrintStream;
import org.apache.commons.lang3.ObjectUtils;
import org.apache.commons.lang3.StringUtils;
import org.fusesource.jansi.Ansi;
import org.fusesource.jansi.Ansi.Color;

/**
 * Denotes a nonempty set of consecutive characters in a source file. Empty ranges are not allowed,
 * since there is no column number that corresponds to a single position.
 */
@AutoValue
public abstract class SourceRange {
  /**
   * Create a range given the positions that surround a sequence of characters.
   *
   * @param start position that comes just before the characters in the range
   * @param end position that comes just after the characters in the range
   */
  static SourceRange create(SourcePosition start, SourcePosition end) {
    if (!start.getSourceFile().equals(end.getSourceFile())) {
      throw new IllegalArgumentException("Positions lie in different files.");
    }
    if (start.equals(end)) {
      throw new IllegalArgumentException("Empty source range.");
    }
    if (start.getOffset() > end.getOffset()) {
      throw new IllegalArgumentException("Start position comes after the end position.");
    }

    return new AutoValue_SourceRange(start, end);
  }

  /** Start position. Comes just before the characters in the range. */
  public abstract SourcePosition getStart();

  /** End position. Comes just after the characters in the range. */
  public abstract SourcePosition getEnd();

  /** Description of where the source file came from. */
  public final String getSourcePath() {
    return getSourceFile().getPath();
  }

  private SourceFile getSourceFile() {
    return getStart().getSourceFile();
  }

  /** Combine two ranges to create a range that spans both. */
  public final SourceRange merge(SourceRange that) {
    final SourcePosition start = ObjectUtils.min(this.getStart(), that.getStart());
    final SourcePosition end = ObjectUtils.max(this.getEnd(), that.getEnd());
    return SourceRange.create(start, end);
  }

  @Override
  public final String toString() {
    final int startColumn = getStart().getColumn();
    final int endColumn = getEnd().getColumn() - 1;

    if (getStart().getLine() == getEnd().getLine()) {
      final String columns = startColumn + "-" + endColumn;
      return getSourcePath() + ":" + getStart().getLine() + ":" + columns;
    } else {
      final String start = "(" + getStart().getLine() + ":" + startColumn + ")";
      final String end = "(" + getEnd().getLine() + ":" + endColumn + ")";
      return getSourcePath() + ":" + start + "-" + end;
    }
  }

  /**
   * Print the relevant portions of the source file and highlight the region that corresponds to
   * this location.
   *
   * @param output where to print the output
   * @param contextLines number of lines before and after the relevant region to display to give
   *     more context to the user
   */
  public void showInSource(PrintStream output, int contextLines) {
    final int startLine = getStart().getLine();
    final int endLine = getEnd().getLine();

    // Number of characters it takes to represent the largest line number.
    final int lineNumberWidth = Integer.toString(endLine).length();

    // True if we are highlighting multiple lines; false otherwise.
    final boolean multiLineMode = startLine != endLine;

    // Print relevant lines
    final int firstLine = Math.max(1, startLine - contextLines);
    final int lastLine = Math.min(getSourceFile().numberOfLines(), endLine + contextLines);
    for (int line = firstLine; line <= lastLine; line++) {
      final String lineNumber = String.format("%" + lineNumberWidth + "d|", line);
      final boolean highlight = startLine <= line && line <= endLine;

      // Print line number
      output.print(lineNumber);

      // Print multiline error indicator
      if (multiLineMode) {
        if (highlight) {
          output.print(Ansi.ansi().fg(Color.RED).a('>').reset());
        } else {
          output.print(" ");
        }
      }

      // Print space between line numbers and line contents
      output.print(" ");

      // Print the actual line
      output.println(getSourceFile().getLine(line));

      // Print error indicator for single-line mode
      if (highlight) {
        final int startColumn = getStart().getColumn();
        final int endColumn = getEnd().getColumn();
        final int highlightStartColumn = lineNumber.length() + 1 + startColumn;
        final int highlightLength = endColumn - startColumn;
        output.print(StringUtils.repeat(' ', highlightStartColumn - 1));
        output.println(
            Ansi.ansi().fg(Color.RED).a(StringUtils.repeat('^', highlightLength)).reset());
      }
    }

    // Make sure there is uniform vertical space after the displayed source code.
    if (multiLineMode || contextLines > 0) {
      // Last line did not have an underline. Add blank line instead.
      output.println();
    }
  }

  /** Similar as {@link #showInSource(PrintStream, int)}, but with context set to a good default. */
  public void showInSource(PrintStream output) {
    final int contextLines = getStart().getLine() == getEnd().getLine() ? 0 : 1;
    showInSource(output, contextLines);
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/parser/SourceRange.kt
package edu.cornell.cs.apl.viaduct.parser

import org.fusesource.jansi.Ansi
import java.io.PrintStream

/**
 * Denotes a nonempty set of consecutive characters in a source file. Empty ranges are not allowed,
 * since there is no column number that corresponds to a single position.
 *
 * @param start position that comes just before the characters in the range
 * @param end position that comes just after the characters in the range
 */
data class SourceRange(private val start: SourcePosition, private val end: SourcePosition) {
    init {
        require(start.sourceFile == end.sourceFile) { "Positions must be in the same file." }
        require(start != end) { "Source range cannot be empty." }
        require(start.offset <= end.offset) { "Start position cannot come after the end position." }
    }

    /** Description of where the source file came from.  */
    val sourcePath: String
        get() = sourceFile.path

    private val sourceFile: SourceFile
        get() = start.sourceFile

    /**
     * Combines two ranges to create a range that spans both.
     *
     * @return The merged range.
     */
    fun merge(that: SourceRange): SourceRange {
        val start = minOf(this.start, that.start)
        val end = maxOf(this.end, that.end)
        return SourceRange(start, end)
    }

    /**
     * Prints the relevant portions of the source file and highlights the region that corresponds to
     * this location.
     *
     * @param output where to print the output
     * @param contextLines number of lines before and after the relevant region to display to give
     * more context to the user
     */
    fun showInSource(
        output: PrintStream,
        contextLines: Int = if (start.line == end.line) 0 else 1
    ) {
        // Number of characters it takes to represent the largest line number.
        val lineNumberWidth = end.line.toString().length

        // True if we are highlighting multiple lines; false otherwise.
        val multiLineMode = start.line != end.line

        // Print relevant lines
        val firstLine = (start.line - contextLines).coerceAtLeast(1)
        val lastLine = (end.line + contextLines).coerceAtMost(sourceFile.numberOfLines)
        for (line in firstLine..lastLine) {
            val lineNumber = String.format("%${lineNumberWidth}d|", line)
            val highlightThisLine = line in start.line..end.line

            // Print line number
            output.print(lineNumber)

            // In multiline mode, mark the entire line as relevant with an indicator
            if (multiLineMode) {
                if (highlightThisLine) {
                    output.print(Ansi.ansi().fg(Ansi.Color.RED).a('>').reset())
                } else {
                    output.print(" ")
                }
            }

            // Print space between line numbers and line contents
            output.print(" ")

            // Print the actual line
            output.println(sourceFile.getLine(line))

            // In single-line mode, underline the relevant portion
            if (!multiLineMode && highlightThisLine) {
                val highlightStartColumn = lineNumber.length + 1 + start.column
                val highlightLength = end.column - start.column
                output.print(" ".repeat(highlightStartColumn - 1))
                output.println(
                    Ansi.ansi().fg(Ansi.Color.RED).a("^".repeat(highlightLength)).reset()
                )
            }
        }

        // Make sure there is uniform vertical space after the displayed source code.
        if (multiLineMode || contextLines > 0) {
            // Last line did not have an underline. Add blank line instead.
            output.println()
        }
    }

    override fun toString(): String {
        val startColumn = start.column
        val endColumn = end.column - 1
        return if (start.line == end.line) {
            "$sourcePath:${start.line}:$startColumn-$endColumn"
        } else {
            val start = "(${start.line}:$startColumn)"
            val end = "(${end.line}:$endColumn)"
            "$sourcePath:$start-$end"
        }
    }
}
